Use WebP for thumbnails and fix seeking on apple browsers

satyamtg opened the pull request:
This allows using WebP for thumbnails with optimization and S3 cache support. Also applies the same CSS tweaks (alongwith webp-hero) and fixes #91 as a result.

This does the following chages -
- thumbnail convertion logic is moved from `processing.py` to `download_video_files_batch()` in `scraper.py ` in the form of a function `process_thumbnail()`.
- templates are updated to use WebP thumbnails
- thumbnails are now optimized and uploaded to cache

rgaudin commented on youtube2zim/scraper.py:
```diff
@@ -585,53 +588,102 @@ def upload_to_cache(self, key, video_path, encoder_version):
+                video_downloaded_from_cache = self.download_from_cache(
+                    s3_key_video, video_path, preset_video.VERSION
+                )
+                thumbnail_downloaded_from_cache = self.download_from_cache(
```
this method has become too large and hard to read with many conditions.

Because neither S3 DL/UL nor youtube-dl save requests for getting the video and the thumbnail together, I suggest we split this into distinct (similar) functions for the video and the thumbnail. It should make it all a lot clearer. Might remove the need for that `record_success_and_log` that I don't like either

satyamtg replied:
Done. Exploded into two methods

rgaudin commented on youtube2zim/scraper.py:
```diff
@@ -699,7 +751,7 @@ def update_metadata(self):
             height=48,
             method="thumbnail",
-            to=self.build_dir.joinpath("favicon.jpg"),
+            dst=self.build_dir.joinpath("favicon.jpg"),
```
hum, so that was broken since moving to new image module I guess

satyamtg replied:
Yup

rgaudin reviewed (changes requested):
Thanks I think we should reorg code a bit as this method is becoming unreadable.

satyamtg commented on youtube2zim/scraper.py:
```diff
@@ -585,53 +588,102 @@ def upload_to_cache(self, key, video_path, encoder_version):
         succeeded = []
         failed = []
 
+        def process_thumbnail(thumbnail_path, preset_thumbnail):
```
Done

satyamtg commented on youtube2zim/scraper.py:
```diff
@@ -585,53 +588,102 @@ def upload_to_cache(self, key, video_path, encoder_version):
             if self.s3_storage:
-                s3_key = f"{self.video_format}/{self.video_quality}/{video_id}"
+                s3_key_video = f"{self.video_format}/{self.video_quality}/{video_id}"
+                s3_key_thumbnail = f"thumbnails/{video_id}"
```
Done. Kept it "high"

satyamtg commented on youtube2zim/scraper.py:
```diff
@@ -585,53 +588,102 @@ def upload_to_cache(self, key, video_path, encoder_version):
         # set preset
-        preset = {"mp4": VideoMp4Low}.get(self.video_format, VideoWebmLow)()
+        preset_video = {"mp4": VideoMp4Low}.get(self.video_format, VideoWebmLow)()
+        preset_thumbnail = WebpMedium()
```
Done

satyamtg commented on CHANGELOG:
```diff
@@ -1,5 +1,6 @@
 2.1.8.dev0
 
+use WebP for thumbnails and fix seeking on Apple browsers
```
Done

satyamtg commented:
> Thanks I think we should reorg code a bit as this method is becoming unreadable.

Added two methods,  `download_video()` and `download_thumbnail()`, to download those seperately
Moved `process_thumbnail()` to `processing.py` module

rgaudin commented on youtube2zim/scraper.py:
```diff
@@ -587,103 +586,106 @@ def upload_to_cache(self, key, video_path, encoder_version):
-                thumbnail_downloaded_from_cache = self.download_from_cache(
-                    s3_key_thumbnail, thumbnail_path, preset_thumbnail.VERSION
-                )
+        if not downloaded_from_cache:
```
Useless block as you return earlier if it's True

rgaudin commented on youtube2zim/scraper.py:
```diff
@@ -587,103 +586,106 @@ def upload_to_cache(self, key, video_path, encoder_version):
-                f"Successfully downloaded and processed assets for video ID - {video_id}"
+                f"Attempting to download thumbnail for {video_id} from cache..."
             )
+            downloaded_from_cache = self.download_from_cache(
```
Can be simplified to 

```py
if self.download_from_cache(s3_key, thumbnail_path, preset.VERSION):
    return True
````

rgaudin commented on youtube2zim/scraper.py:
```diff
@@ -587,103 +586,106 @@ def upload_to_cache(self, key, video_path, encoder_version):
 
-        def record_success_and_log(video_id):
-            succeeded.append(video_id)
+        downloaded_from_cache = False
```
useless (see bellow)

rgaudin commented on youtube2zim/scraper.py:
```diff
@@ -587,103 +586,106 @@ def upload_to_cache(self, key, video_path, encoder_version):
+                logger.debug(exc)
+                return False
+            else:  # upload to cache only if everything went well
+                if self.s3_storage and not downloaded_from_cache:
```
Again, `downloaded_from_cache` would always be False so should be removed

rgaudin reviewed (changes requested):
Good. Can be simplified a bit (see comments in thumbnail and apply to video as well)

satyamtg commented:
> Good. Can be simplified a bit (see comments in thumbnail and apply to video as well)

Hm. I wonder why I didn't saw those. Fixed them anyways.

rgaudin reviewed (changes requested):
<img width="582" alt="Screen Shot 2020-09-15 at 10 25 23" src="https://user-images.githubusercontent.com/57929/93199325-c5b10580-f73d-11ea-9125-fa3edcdc61d7.png">

There are multiple left over files (broken JPEG) regarding thumbnails… what's going on?

satyamtg commented on youtube2zim/scraper.py:
```diff
@@ -640,39 +635,34 @@ def download_thumbnail(self, video_id, options):
+            options_copy["skip_download"] = True
+            with youtube_dl.YoutubeDL(options_copy) as ydl:
+                ydl.download([video_id])
+            process_thumbnail(thumbnail_path, preset)
```
Fixed that

satyamtg commented:
> <img alt="Screen Shot 2020-09-15 at 10 25 23" width="582" src="https://user-images.githubusercontent.com/57929/93199325-c5b10580-f73d-11ea-9125-fa3edcdc61d7.png">
> 
> There are multiple left over files (broken JPEG) regarding thumbnails… what's going on?

This was apparently due to the configuration option we pass to `youtube_dl`, i.e. `write_all_thumbnails`. Setting it to `False` did the trick.

satyamtg commented:
I have added a function to download subtitles. Also bumped scraperlib requirement to 1.3.4 because it requires https://github.com/openzim/python_scraperlib/pull/65 to correctly handle thumbnails which are often downloaded with wrong extensions by youtube_dl.

rgaudin reviewed (approved):
OK, I have some out-of-scope concerns for which I'm opening a separate ticket.
